crates/mcp-client/src/transport/stdio.rs

```diff
@@ -16,20 +16,59 @@ pub struct StdioActor {
     receiver: mpsc::Receiver<TransportMessage>,
     pending_requests: Arc<PendingRequests>,
     _process: Child, // we store the process to keep it alive
+    error_sender: mpsc::Sender<Error>,
     stdin: ChildStdin,
     stdout: ChildStdout,
+    stderr: ChildStderr,
 }
 
 impl StdioActor {
-    pub async fn run(self) {
-        tokio::join!(
-            Self::handle_incoming_messages(self.stdout, self.pending_requests.clone()),
-            Self::handle_outgoing_messages(
-                self.receiver,
-                self.stdin,
-                self.pending_requests.clone()
-            )
+    pub async fn run(mut self) {
+        use tokio::pin;
+
+        let incoming = Self::handle_incoming_messages(self.stdout, self.pending_requests.clone());
+        let outgoing = Self::handle_outgoing_messages(
+            self.receiver,
+            self.stdin,
+            self.pending_requests.clone(),
         );
+
+        // take ownership of futures for tokio::select
+        pin!(incoming);
+        pin!(outgoing);
+
+        // Use select! to wait for either I/O completion or process exit
+        tokio::select! {
+            result = &mut incoming => {
+                tracing::debug!("Stdin handler completed: {:?}", result);
+            }
+            result = &mut outgoing => {
+                tracing::debug!("Stdout handler completed: {:?}", result);
+            }
+            // capture the status so we don't need to wait for a timeout
+            status = self._process.wait() => {
```

Comment:
are we supposed to sth with the status here? looks like we're just logging it.

Reply:
i mainly kept it for logging for debug purposes, i suppose we could skip the `stderr.read_to_end` for a `0` exit code to be slightly more efficient?